### PR TITLE
ECER-2256 - BE: Create endpoint to send messages

### PR DESCRIPTION
**Pull request by qsl-deepak**
---
name: ECER-2256 - BE: Create endpoint to send messages
about: BE: Create endpoint to send messages
---

## Title
ECER-2256 - BE: Create endpoint to send messages

## Description

- Added new endpoint for sending messages as portal user (registrant)
- Added handler and repository layer code
- Added/updated prop mappings in all layers
- Added Automated test for sending messages.
- Added Sanitization on the text message coming from the portal user

## Related Jira Issue(s)

- ECER-2256
- ECER-1252

## Checklist
- [x] I have tested these changes locally.
- [x] Changes to backend endpoints are covered by passing integration tests.
- [ ] I have added or updated the necessary documentation.

### CODE REVIEW

File: src/ECER.Managers.Registry.Contract/Communications/Contract.cs [hunk 1]
```diff
@@ -22,6 +22,19 @@ public record UserCommunicationQuery : IRequest<CommunicationsQueryResults>
 }
 public record CommunicationsQueryResults(IEnumerable<Communication> Items);
 
+public record SendMessageCommand(Communication communication, string userId) : IRequest<SendMessageResult>;
+
+public class SendMessageResult
+{
+  public string? CommunicationId { get; set; }
```
Comment (SoLetsDev) [inline src/ECER.Managers.Registry.Contract/Communications/Contract.cs:29]: should CommunicationId always be present for a post?

File: src/ECER.Resources.Accounts/Communications/CommunicationRepository.cs [hunk 2]
```diff
@@ -64,4 +65,45 @@ public async Task<string> MarkAsSeen(string communicationId, CancellationToken c
...
+    var existingCommunication = context.ecer_CommunicationSet.FirstOrDefault(
+      d => d.ecer_CommunicationId == Guid.Parse(communication.Id!)
+      );
+    if (existingCommunication == null)
+    {
+      throw new InvalidOperationException($"Communication '{communication.Id}' not found");
+    }
+    else
+    {
+      var application = context.ecer_ApplicationSet.FirstOrDefault(a => a.ecer_ApplicationId == existingCommunication.ecer_Applicationid.Id && a.ecer_Applicantid.Id == Guid.Parse(userId));
```
Comment (farzadnadiri) [inline src/ECER.Resources.Accounts/Communications/CommunicationRepository.cs:81]: consider using Single instead, FirstOrDefault can cause bugs

File: src/ECER.Managers.Registry.Contract/Communications/Contract.cs (hunk 1 above)
Comment (qsl-deepak) [reply src/ECER.Managers.Registry.Contract/Communications/Contract.cs:29]: @SoLetsDev - in the SendMessageResult class -> is just for returning newly added communication(message) ID back as response...

Comment (SoLetsDev) [reply src/ECER.Managers.Registry.Contract/Communications/Contract.cs:29]: oh I get it, the result can be a successful or a failure... which might not have a communicationId

File: src/ECER.Resources.Accounts/Communications/CommunicationRepository.cs (hunk 2 above)
Comment (qsl-deepak) [reply src/ECER.Resources.Accounts/Communications/CommunicationRepository.cs]: good point @farzadnadiri, Fixed...

File: src/ECER.Managers.Registry.Contract/Communications/Contract.cs [hunk 3]
```diff
@@ -22,6 +22,19 @@ public record UserCommunicationQuery : IRequest<CommunicationsQueryResults>
...
+  public bool IsSuccess { get; set; }
+  public string? ErrorMessage { get; set; }
+
+  public static SendMessageResult Success() => new SendMessageResult { IsSuccess = true };
+
+  public static SendMessageResult Failure(string message) => new SendMessageResult { IsSuccess = false, ErrorMessage = message };
+}
```
Comment (SoLetsDev) [inline src/ECER.Managers.Registry.Contract/Communications/Contract.cs:36]: lets remove the two methods if we aren't using them :)

Comment (qsl-deepak) [reply src/ECER.Managers.Registry.Contract/Communications/Contract.cs:32]: Done..

**Review by SoLetsDev — APPROVED** [review]

**Review by farzadnadiri — APPROVED** [review]

**Pull request merged**